impl `RecordHeader` and perform PUT validations

RolandSherwin opened the pull request:
Closes #318 and #323 

- Implement `RecordHeader` to make sure we are not overwriting any `Record`
- Perform validations on PUT when we get them, comparing them with the local copy.
- Move `replication logic` to the node level. This allows us to perform validation on the `ReplicationData` as well.
- Trim `RecordHeader` and return only the requested item during `GET`
- Sending `Request` without awaiting a `Response` has been flushed out.

RolandSherwin committed:
feat(kad): impl `RecordHeader` to store the record kind

RolandSherwin committed:
feat(network): validate and store `Record`

RolandSherwin committed:
feat(network): store `Chunk` along with `PaymentProof`

RolandSherwin committed:
feat(node): perform proper validations on PUT

RolandSherwin committed:
feat(network): validate and store `ReplicatedData`

RolandSherwin committed:
fix(network): Send `Request` without awaiting for `Response`
- The `Response` is instead handled using the commong `handle_response`
  function instead of awaiting for the response from the peer at the
  call site.

RolandSherwin committed:
feat(network): validate `Record` on GET

RolandSherwin committed:
chore(docs): add more docs and comments

joshuef commented on sn_protocol/src/storage/header.rs:
```diff
@@ -0,0 +1,59 @@
+}
+
+impl RecordHeader {
+    // Bincode serializes enums with unit variants as a u32, not a u8, so it would take up 4 bytes.
```
I don't think we can use bincode here. The protocol should be language agnostic and `bincode` is rust specific

RolandSherwin replied:
Updated the PR to use `rmp_serde` instead.

RolandSherwin committed:
fix(network): use `rmp_serde` for `RecordHeader` ser/de

RolandSherwin committed:
chore(workflow): fix data replication script

RolandSherwin committed:
fix(node): store parent tx along with `SignedSpend`

RolandSherwin committed:
fix(record_header): encode unit enum as u32
- rmp_serde serializes unit variant as it pleases, thus the encode size
  is not fixed.

grumbach commented on sn_networking/src/lib.rs:
```diff
@@ -520,51 +548,127 @@ impl Network {
             .await)
     }
 
-    /// Get `key` from our Storage
-    pub async fn get_provided_data(&self, key: RecordKey) -> Result<Result<Vec<u8>>> {
+    /// Returns the list of keys that are within the provided distance to the target
+    pub async fn get_record_keys_closest_to_target(
+        &self,
+        target: &NetworkAddress,
+        distance: Distance,
+    ) -> Result<Vec<RecordKey>> {
         let (sender, receiver) = oneshot::channel();
-        self.send_swarm_cmd(SwarmCmd::GetData { key, sender })
-            .await?;
+        self.send_swarm_cmd(SwarmCmd::GetRecordKeysClosestToTarget {
+            key: target.clone(),
+            distance,
+            sender,
+        })
+        .await?;
 
         receiver
             .await
             .map_err(|_e| Error::InternalMsgChannelDropped)
     }
 
-    /// Get `ReplicatedData` from local Storage
-    pub async fn get_replicated_data(
-        &self,
-        address: NetworkAddress,
-    ) -> Result<Result<QueryResponse>> {
+    /// Get the Record from the network
+    pub async fn get_record_from_network(&self, key: RecordKey) -> Result<Record> {
```
Excellent renaming! Was always confused with the previous name!

grumbach commented on sn_node/src/api.rs:
```diff
@@ -188,131 +203,151 @@ impl Node {
+                        // Put validations make sure that we have >= 1 spends and with the same
+                        // dbc_id
+                        let dbc_addr =
+                            DbcAddress::from_dbc_id(spends_with_parent[0].signed_spend.dbc_id());
```
`[0]` is unsafe, might panic if the vec is empty

grumbach commented on sn_node/src/put_validation.rs:
```diff
@@ -0,0 +1,433 @@
+        })?;
+
+        // Notify the sender of any double spend
+        if spends_with_parent.len() > 1 {
```
Here we don't check if it is the same DBC, are we sure that it's not, aka are we sure it is a double spend and not twice the same dbc?

grumbach commented on sn_node/src/put_validation.rs:
```diff
@@ -0,0 +1,433 @@
+                }
+                // Since spends_with_parent is not empty and they contain the same DbcId (RecordKeys are
+                // the same), get the DbcId
+                let dbc_id = *spends_with_parent[0].signed_spend.dbc_id();
```
Here I would rather use a match instead of relying on `[0]`which is a dangerous pattern

grumbach commented on sn_node/src/put_validation.rs:
```diff
@@ -0,0 +1,433 @@
+    /// - If the SpendWithParent for the provided DbcId is present locally, check for new spends by
+    /// comparing it with the local copy.
+    /// - If incoming spends_with_parent.len() > 1, aggregate store them directly as they are a double spent.
+    /// - If incoming spends_with_parent.len() == 1, then check for parent_inputs and the closet(dbc_id)
```
`closet` typo

grumbach commented on sn_node/src/put_validation.rs:
```diff
@@ -0,0 +1,433 @@
+        if present_locally {
+            debug!("DbcSpend with DbcId {dbc_id:?} already exists, checking if it's the same spend/double spend",);
+            // fetch the locally stored record; should be present
+            //
```
Typo?

grumbach commented on sn_node/src/put_validation.rs:
```diff
@@ -0,0 +1,433 @@
+            debug!(
+                "Received a single SpendWithParent, verifying the parent and checking for double spend"
+            );
+            let spends_with_parent = spends_with_parent.remove(0);
```
Here a match would be less dangerous than check then remove

grumbach commented on sn_node/src/replication.rs:
```diff
@@ -0,0 +1,207 @@
+            return Ok(());
+        };
+        let distance_bar =
+            NetworkAddress::from_peer(sorted_peers[CLOSE_GROUP_SIZE]).distance(&our_address);
```
`sorted_peers[CLOSE_GROUP_SIZE]` from a `Vec<PeerId>` is unsafe

grumbach commented on sn_node/src/replication.rs:
```diff
@@ -0,0 +1,207 @@
+            return Ok(());
+        };
+
+        let distance_bar = NetworkAddress::from_peer(sorted_peers[CLOSE_GROUP_SIZE])
```
Same as above, I suggest we use another way like:
- pattern matching
- get() -> Option and check it

grumbach commented on sn_node/src/replication.rs:
```diff
@@ -0,0 +1,207 @@
+            };
+
+            // Only carry out replication when self within REPLICATION_RANGE
+            let replicate_range = NetworkAddress::from_peer(closest_peers[REPLICATION_RANGE]);
```
same here, since we use this one a lot, maybe we could get it out in a variable and reuse it everywhere we need?

grumbach commented on sn_node/src/spends.rs:
```diff
@@ -0,0 +1,229 @@
+        )));
+    }
+
+    // get the parent spends form the network
```
```suggestion
    // get the parent spends from the network
```

grumbach commented on sn_protocol/src/storage/spend.rs:
```diff
@@ -0,0 +1,30 @@
+
+/// The `SignedSpend` along with its corresponding `ParentTx`
+#[derive(custom_debug::Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SpendWithParent {
```
With https://github.com/maidsafe/sn_dbc/pull/278 in, we should be able to get rid of this

grumbach replied:
Maybe in a follow up PR?

grumbach reviewed (commented):
Big one, and a good one! Left a couple comments mostly about code safety, but generally looking good!

RolandSherwin committed:
fix(network): use safe operations when dealing with Vec

grumbach reviewed (approved):
LGTM 🎉
